chain-api/src/types/TokenBalance.spec.ts

```diff
@@ -112,11 +120,145 @@ describe("fungible", () => {
+    balance.ensureCanSubtractQuantity(subtractTotal, Date.now()).subtract();
+
+    // Then
+    expect(balance.getQuantityTotal()).toEqual(initialTotal.minus(subtractTotal));
+    expect(balance.getLockedQuantityTotal(Date.now())).toEqual(new BigNumber("0"));
+    expect(balance.getSpendableQuantityTotal(Date.now())).toEqual(initialTotal.minus(subtractTotal));
+  });
+
+  it("should fail to unlock a TokenHold Quantity with mismatched name", () => {
+    // Given
+    const initialTotal = new BigNumber("10");
+    const lockedTotal = new BigNumber("10");
+
+    const balance = emptyBalance();
+    const nameForHold = "client|someone-else";
+    const nameForUnlockAttempt = undefined;
+    const hold = createHold(TokenInstance.FUNGIBLE_TOKEN_INSTANCE, 0, lockedTotal, nameForHold);
+
+    balance.ensureCanAddQuantity(initialTotal).add();
+    balance.ensureCanLockQuantity(hold).lock();
+
+    const tokenClassKey = TokenClassKey.toStringKey({ ...balance });
+    // When
+    const error = () =>
+      balance.ensureCanUnlockQuantity(lockedTotal, Date.now(), nameForUnlockAttempt).unlock();
+
+    // Then
+    expect(error).toThrow(
+      `Failed to unlock quantity ${lockedTotal} of Fungible token ${tokenClassKey} ` +
+        `for TokenHold.name = ${nameForUnlockAttempt}.`
+    );
+  });
+
+  it("should unlock quantity by using TokenHolds that will expire soonest", () => {
+    // Given
+    const initialTotal = new BigNumber("10");
+    const quantityLockedPerHold = new BigNumber("1");
+
+    const noExpirationTimestamp = 0;
+    const expiresIn30MinutesTimestamp = Date.now() + 1000 * 60 * 30;
+    const expiresIn30DaysTimestamp = Date.now() + 1000 * 60 * 60 * 24 * 30;
+
+    const balance = emptyBalance();
+    const holds = [
+      createHold(
+        TokenInstance.FUNGIBLE_TOKEN_INSTANCE,
+        noExpirationTimestamp,
+        quantityLockedPerHold,
+        undefined
+      ),
+      createHold(
+        TokenInstance.FUNGIBLE_TOKEN_INSTANCE,
+        expiresIn30DaysTimestamp,
+        quantityLockedPerHold,
+        undefined
+      ),
+      createHold(
+        TokenInstance.FUNGIBLE_TOKEN_INSTANCE,
+        expiresIn30MinutesTimestamp,
+        quantityLockedPerHold,
+        undefined
+      )
+    ];
+
+    balance.ensureCanAddQuantity(initialTotal).add();
+
+    holds.forEach((hold) => {
+      balance.ensureCanLockQuantity(hold).lock();
+    });
+
+    const lockedTotal = quantityLockedPerHold.times(holds.length);
+    // When
+    balance.ensureCanUnlockQuantity(quantityLockedPerHold, Date.now()).unlock();
+    const unexpiredLockedHolds = balance.getUnexpiredLockedHolds(Date.now());
+
+    // Then
+    expect(balance.getLockedQuantityTotal(Date.now())).toEqual(lockedTotal.minus(quantityLockedPerHold));
+    expect(unexpiredLockedHolds.length).toBe(2);
+    expect(unexpiredLockedHolds[0].expires).toBe(expiresIn30DaysTimestamp);
+    expect(unexpiredLockedHolds[1].expires).toBe(noExpirationTimestamp);
+  });
```

Comment:
great tests! I think we can also add a test to verify partial unlock passes, like add two locks with quantity = 5, and unlock 7

Reply:
I'm glad you recommended this test case, because while implementing I realized I'd failed to handle that scenario yesterday. Fixed in latest commit.